### PR TITLE
chore: Divvy up vector-core

### LINKED ISSUE #7407: Divvy up `vector-core` into crates
As a part of #7400 we've discussed that `vector-core` should be split up into crates with `buffers` being an isolated crate.

### PR DESCRIPTION
**Pull request by blt**
This commit extracts 'buffers' into its own crate inside the vector-core
workspace and slims dependencies in both crates. Calling code is unchanged as
the top-level crate re-exports buffers. I intend to move tests and benchmarks of
buffers into the sub-crate in a subsequent PR.

Resolves #7407

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

File: lib/vector-core/buffers/src/lib.rs [hunk 1]
```diff
@@ -178,7 +193,7 @@ impl<T, S: Sink<T> + Unpin> Sink<T> for DropWhenFull<S> {
                 *this.drop = true;
                 Poll::Ready(Ok(()))
             }
-            error => error,
+            error @ std::task::Poll::Ready(..) => error,
```
Comment (tobz) [inline lib/vector-core/buffers/src/lib.rs:196]: Is this to be more explicit that we're doing passthrough in the error case?

Comment (blt) [reply lib/vector-core/buffers/src/lib.rs:196]: Yep, this is a clippy pedantic ding, specifically https://rust-lang.github.io/rust-clippy/master/#wildcard_enum_match_arm. We'd flagged it off in the top-level vector-core but here, at least, I buy being explicit.

File: src/sources/syslog.rs [hunk 2]
```diff
@@ -568,7 +567,7 @@ mod test {
           "#,
         )
         .unwrap();
-        assert!(config.mode.is_tcp());
+        assert!(matches!(config.mode, Mode::Tcp { .. }));
```
Comment (bruceg) [inline src/sources/syslog.rs:570]: If this is all it takes to avoid another `derive` procedural macro, then good riddance.

**Review by bruceg — APPROVED** [review]

File: src/sources/syslog.rs (hunk 2 above)
Comment (blt) [reply src/sources/syslog.rs:570]: Yeah, I stumbled backward into this realization and was very pleased.

**Pull request merged**